url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/client/SOCDiscardDialog.java

```diff
@@ -0,0 +1,351 @@
+            discThese.setBounds(getInsets().left, getInsets().top + 20 + space + 20 + space + sqwidth + space, 100, 20);
+        }
+        catch (NullPointerException e) {}
+
+        keepY = getInsets().top + 20 + space + 20 + space;
+        discY = keepY + sqwidth + space + 20 + space;
+
+        try
+        {
+            for (int i = 0; i < 5; i++)
+            {
+                keep[i].setSize(sqwidth, sqwidth);
+                keep[i].setLocation(x + sqspace + (i * (sqspace + sqwidth)), keepY);
+                disc[i].setSize(sqwidth, sqwidth);
+                disc[i].setLocation(x + sqspace + (i * (sqspace + sqwidth)), discY);
+            }
+        }
+        catch (NullPointerException e) {}
+    }
+
+    /**
+     * DOCUMENT ME!
+     *
+     * @param e DOCUMENT ME!
+     */
+    public void actionPerformed(ActionEvent e)
+    {
+        try {
+        Object target = e.getSource();
+
+        if (target == discardBut)
+        {
+            SOCResourceSet rsrcs = new SOCResourceSet(disc[0].getIntValue(), disc[1].getIntValue(), disc[2].getIntValue(), disc[3].getIntValue(), disc[4].getIntValue(), 0);
+
+            if (rsrcs.getTotal() == numDiscards)
+            {
+                playerInterface.getClient().discard(playerInterface.getGame(), rsrcs);
+                dispose();
+            }
+        }
+        } catch (Throwable th) {
+            playerInterface.chatPrintStackTrace(th);
+        }
+    }
+
+    /**
+     * DOCUMENT ME!
+     *
+     * @param e DOCUMENT ME!
+     */
+    public void mouseEntered(MouseEvent e)
+    {
+        ;
+    }
+
+    /**
+     * DOCUMENT ME!
+     *
+     * @param e DOCUMENT ME!
+     */
+    public void mouseExited(MouseEvent e)
+    {
+        ;
+    }
+
+    /**
+     * DOCUMENT ME!
+     *
+     * @param e DOCUMENT ME!
+     */
+    public void mouseClicked(MouseEvent e)
+    {
+        ;
+    }
+
+    /**
+     * DOCUMENT ME!
+     *
+     * @param e DOCUMENT ME!
+     */
+    public void mouseReleased(MouseEvent e)
+    {
+        ;
+    }
+
+    /**
+     * When a resource's colorsquare is clicked, add/remove 1
+     * from the resource totals as requested; update {@link #discardBut}.
+     */
+    public void mousePressed(MouseEvent e)
+    {
+        try {
+        Object target = e.getSource();
+        boolean wantsRepaint = false;
+
+        for (int i = 0; i < 5; i++)
+        {
+            if ((target == keep[i]) && (disc[i].getIntValue() > 0))
+            {
+                keep[i].addValue(1);
+                disc[i].subtractValue(1);
+                --numChosen;
+                if (numChosen == (numDiscards-1))
+                {
+                    discardBut.disable();  // Count un-reached (too few)
+                    wantsRepaint = true;
+                }
+                else if (numChosen == numDiscards)
+                {
+                    discardBut.enable();   // Exact count reached
+                    wantsRepaint = true;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 312:321, 329:338.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.